### ISSUE TITLE
Handling Massive Images

### ISSUE THREAD

**Issue by EyeAndTea** [dlemstra/Magick.NET#222]
Good day.

I apologize for opening an issue to ask a question, if this is not the expected conduct here. My understanding is that one can stream a massive image converting it to the mpc format using imagemagick, streaming it without having it load the entire image into memory. One can then use the mpc file for further processing without having to load fully the mpc file either.

1- Is this possible with Magick.NET?
2- If not, is it possible to crop into bits a very large image without loading it fully into memory? 

I am aware of the existence of NetVips, but I am targeting framework 4.0.3, and have special requirements. Magick.NET appears to be already creating temporary files, which I am assuming are mpc, under certain conditions, and hence why I wondered.

**Comment by dlemstra** [comment]
You can limit the memory that is being used with the `ResourceLimits` class. But you will still need to fully decode your image. We don't support only loading part of the image.

**Comment by EyeAndTea** [comment]
Thank you for your time. Apologies, perhaps I am not being clear. If I may, let me try again.

A1- I am understanding that straight streaming from image to mpc on initial load is not possible, unlike with imagemagick itself if I am not wrong.
A2- I am understanding that working on the mpc files themselves is also not possible, unlike with imagemagick itself.

If both my understandings are correct, and if what you say is also correct, there might exists a bug. Please note:

B1- Limiting the memory using ResourceLimits has only caused an exception to be thrown when trying to work on a large bitmap. No temporary files were created. However when a file stream was used instead of a string for the image's full path, a temporary file does get created. That is all I know regarding this.
B2- With jpeg, the image is always decoded to temporaries, keeping memory usage low, which is part of the behavior that I want.

**Comment by dlemstra** [comment]
You can convert an image to an mpc file when you use that as the output format: `image.Write("file.mpc")`. The things that you can do with ImageMagick should also be possible with Magick.NET. Do you have trouble creating an mpc file?

**Comment by EyeAndTea** [comment]
Not without loading the whole original image into memory. Please give me your attention, there are a number of issues here. 

1- **The original question** was about converting an image, that can be trivially streamed, like a bitmap, into mpc without loading the whole image into memory, which should be possible with ImageMagick if my understanding is correct. However, no, I am unable to convert into mpc using Magick.Net without loading the whole thing into memory. 

2- Part of **the original question** also was that I would like to work with mpc files without loading them fully into memory, assuming I manage to create them. Judging by Magick.Net own behavior with the temporary files, and assuming they are mpc files, then this should be possible, but whether this facility is available or not to me, the client coder, is the question. A question that is not important if the answer to '1' is, no it is not possible.

3- After our exchange it occurred to me that there might be **a bug**, which is somewhat related to our discussion, but **is a different matter**. The temporaries are not always created. Loading a large bitmap does not cause a temporary to be created to meet memory limits, and instead Magick.Net throws an exception about its limits reached. The limits set on ResourceLimits. If loading the large bitmap from a file stream, temporaries are created, and is very slow, but that is another matter.

**Comment by dlemstra** [comment]
You cannot create an MPC file without loading the whole image. This is because the image needs to be decoded completely. But when ImageMagick runs out of memory (because you try to load a massive image) it will swap to disk and put the pixels into an temporary file on disk. This will happen automatically.

**Comment by EyeAndTea** [comment]
It seems we are unable to tune to the same frequency. Or perhaps I am not understanding something. Whatever it is, and putting ImageMagick aside, a bitmap in general should be processable for such a task without needing to load it whole into memory, which is expected to be not the same for something like jpeg, but I do not know ennough about the encoding scheme of jpeg.

However, and for the benefit of future readers, please refer to:
- http://www.imagemagick.org/Usage/files/#massive


That is all from me on the matter. Thank you again for your time.
